clientv3/balancer.go

```diff
@@ -26,28 +26,66 @@ import (
 // simpleBalancer does the bare minimum to expose multiple eps
 // to the grpc reconnection code path
 type simpleBalancer struct {
-	// eps are the client's endpoints stripped of any URL scheme
-	eps     []string
-	ch      chan []grpc.Address
+	// addrs are the client's endpoints for grpc
+	addrs []grpc.Address
+
+	ch chan []grpc.Address
+
+	// upc receives an async message whenever up is posted
+	upc chan struct{}
+
+	// mu protects upEps and numGets
+	mu sync.Mutex
+	// upEps contains the endpoints grpc says are up
+	upEps map[string]struct{}
+	// numGets chooses the next endpoint
 	numGets uint32
 }
 
-func newSimpleBalancer(eps []string) grpc.Balancer {
+func newSimpleBalancer(eps []string) *simpleBalancer {
 	ch := make(chan []grpc.Address, 1)
 	addrs := make([]grpc.Address, len(eps))
 	for i := range eps {
 		addrs[i].Addr = getHost(eps[i])
 	}
 	ch <- addrs
-	return &simpleBalancer{eps: eps, ch: ch}
+	return &simpleBalancer{
+		addrs: addrs,
+		ch:    ch,
+		upc:   make(chan struct{}, 1),
+		upEps: make(map[string]struct{}),
+	}
+}
+
+func (b *simpleBalancer) Start(target string) error { return nil }
+
+func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
+	b.mu.Lock()
+	b.upEps[addr.Addr] = struct{}{}
+	b.mu.Unlock()
+	// notify client that a connection is up
+	select {
+	case b.upc <- struct{}{}:
+	default:
+	}
+	return func(error) {
+		b.mu.Lock()
+		delete(b.upEps, addr.Addr)
+		b.mu.Unlock()
+	}
 }
 
-func (b *simpleBalancer) Start(target string) error        { return nil }
-func (b *simpleBalancer) Up(addr grpc.Address) func(error) { return func(error) {} }
 func (b *simpleBalancer) Get(ctx context.Context, opts grpc.BalancerGetOptions) (grpc.Address, func(), error) {
-	v := atomic.AddUint32(&b.numGets, 1)
-	ep := b.eps[v%uint32(len(b.eps))]
-	return grpc.Address{Addr: getHost(ep)}, func() {}, nil
+	b.mu.Lock()
```

Comment:
I think most of that complication is from making it as general as possible. It's safe to assume FailFast is false so there's no need to implement the suggested convoluted blocking logic.